### PR TITLE
frontend part of precision switching

### PR DESCRIPTION
**Pull request by dbochkov-flexcompute**

### CODE REVIEW

**Comment by tylerflex** [comment]
Is the user able to toggle double precision through
```py
tidy3d.constants.TIDY3D_DOUBLE_PRECISION = True
```
and does it propagate to the other areas of the front end?

We could also consider adding this field to the [config](https://github.com/flexcompute/tidy3d/blob/develop/tidy3d/config.py#L16-L59).

**Comment by dbochkov-flexcompute** [comment]
ah, sorry, forgot to copy some of the description from the backend PR. The way I implemented it, `TIDY3D_DOUBLE_PRECISION` is set automatically to either `False` or `True` during the build stage and is not supposed to change after that (if you do change it, that will only affect the python interface but not the solver itself). Definitely would appreciate any feedback on whether this implementation makes sense. If it seems necessary, we could also think about adding an "on-the-fly" capability to switch precision without rebuilding

File: tidy3d/constants.py [hunk 1]
```diff
@@ -13,20 +13,22 @@
...
+# floating point types and precisions
+TIDY3D_DOUBLE_PRECISION = False
+fp_type = np.float64 if TIDY3D_DOUBLE_PRECISION else np.float32
+cfp_type = np.complex128 if TIDY3D_DOUBLE_PRECISION else np.complex64
+fp_eps = np.finfo(fp_type).eps
+
 # fundamental constants
-EPSILON_0 = np.float32(8.85418782e-18)
-MU_0 = np.float32(1.25663706e-12)
+EPSILON_0 = fp_type(8.85418782e-18)
```
Comment (momchil-flex) [inline tidy3d/constants.py:23]: I think we can just remove the conversion here. Some of the explicit conversions to float came from the very early days in the solver, when I was kinda trying to make everything relevant to the solver single precision.

File: tidy3d/constants.py [hunk 2]
```diff
@@ -13,20 +13,22 @@
 
 import numpy as np
 
+# floating point types and precisions
+TIDY3D_DOUBLE_PRECISION = False
+fp_type = np.float64 if TIDY3D_DOUBLE_PRECISION else np.float32
+cfp_type = np.complex128 if TIDY3D_DOUBLE_PRECISION else np.complex64
+fp_eps = np.finfo(fp_type).eps
```
Comment (momchil-flex) [inline tidy3d/constants.py:20]: The frontend uses double precision (or more precisely, whatever precision numpy installation you have, which should almost always be double) almost everywhere. We have been using the single precision epsilon as a shorthand for "a small number" in a few places, even when the operations involved are actually double precision. In that sense I am not sure that using the actual precision epsilon would be sufficient in those places, and maybe we need to replace the current usage of `fp_eps` in the code with some `SMALL_NUMBER = 1e-8` type of thing.

**Review by momchil-flex — COMMENTED** [review]
I think that we should not be modifying anything on the frontend for now, but instead keep the precision switching for internal purposes only for the time being. The `TIDY3D_DOUBLE_PRECISION` global can be moved somewhere in the backend, but we should also discuss if that's the general approach we want to take.

**Comment by tylerflex** [comment]
> I think that we should not be modifying anything on the frontend for now, but instead keep the precision switching for internal purposes only for the time being. The TIDY3D_DOUBLE_PRECISION global can be moved somewhere in the backend, but we should also discuss if that's the general approach we want to take.

Yea I think I agree, maybe the precision should only come in on the front end if it is an optional `Field()` ie in the `ModeSolver` or eventually `Simulation`.  (Basically if it's something the user can change). Otherwise, it should probably be a backend parameter.

**Comment by dbochkov-flexcompute** [comment]
Sounds good to me! Should I just reduce this PR to removing `np.float32` for `EPSILON_0` and `MU_0` in `constants.py`?

**Comment by momchil-flex** [comment]
I think for now this is sufficient for the forntend yeah.

**Pull request merged**